stko/optimizers/xtb.py

```diff
@@ -496,3 +514,633 @@ def optimize(self, mol):
+        ----------
+        xyz : :class:`str`
+            The name of the input structure ``.xyz`` file.
+
+        out_file : :class:`str`
+            The name of output file with xTB results.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        # Modify the memory limit.
+        if self._unlimited_memory:
+            memory = 'ulimit -s unlimited ;'
+        else:
+            memory = ''
+
+        # Set optimization level and type.
+        optimization = f'--opt {self._opt_level}'
+
+        cmd = (
+            f'{memory} {self._xtb_path} {xyz} '
+            f'--gfnff '
+            f'{optimization} --parallel {self._num_cores} '
+            f'--chrg {self._charge} '
+        )
+
+        with open(out_file, 'w') as f:
+            # Note that sp.call will hold the program until completion
+            # of the calculation.
+            sp.call(
+                cmd,
+                stdin=sp.PIPE,
+                stdout=f,
+                stderr=sp.PIPE,
+                # Shell is required to run complex arguments.
+                shell=True
+            )
+
+    def _run_optimization(self, mol):
+        """
+        Run loop of optimizations on `mol` using xTB.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        mol : :class:`.Molecule`
+            The optimized molecule.
+
+        opt_complete : :class:`bool`
+            Returns ``True`` if the calculation is complete and
+            ``False`` if the calculation is incomplete.
+
+        """
+
+        xyz = f'input_structure_ff.xyz'
+        out_file = f'optimization_ff.output'
+        mol.write(xyz)
+        self._run_xtb(xyz=xyz, out_file=out_file)
+
+        # Check if the optimization is complete.
+        output_xyz = 'xtbopt.xyz'
+        opt_complete = self._is_complete(out_file)
+        mol = mol.with_structure_from_file(output_xyz)
+
+        return mol, opt_complete
+
+    def optimize(self, mol):
+        """
+        Optimize `mol`.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        mol : :class:`.Molecule`
+            The optimized molecule.
+
+        """
+
+        if self._output_dir is None:
+            output_dir = str(uuid.uuid4().int)
+        else:
+            output_dir = self._output_dir
+        output_dir = os.path.abspath(output_dir)
+
+        if os.path.exists(output_dir):
+            shutil.rmtree(output_dir)
+
+        os.mkdir(output_dir)
+        init_dir = os.getcwd()
+        os.chdir(output_dir)
+
+        try:
+            mol, complete = self._run_optimization(mol)
+        finally:
+            os.chdir(init_dir)
+
+        if not complete:
+            logging.warning(f'Optimization is incomplete for {mol}.')
+
+        return mol
+
+
+class XTBFFCREST(XTB):
```

Comment:
This class seems to use a very similar approach to the above class. I think we could think about inheriting from a GFN class in the future.

Reply:
I guess in the future. But each method of these classes are necessarily unique